synapse/cryotank.py

```diff
@@ -571,3 +588,609 @@ def init(self, name, conf=None, timeout=None):
+            raise s_exc.TimeOut()
+
+    return wrap
+
+class CryoTankIndexer:
+    '''
+    Manages indexing of a single cryotank's records
+
+    This implements a lazy indexer that indexes a cryotank in a separate thread.
+
+    Cryotank entries are msgpack-encoded json-compatible dictionaries with arbitrary nesting.  An index consists of a
+    property name, one or more datapaths (i.e. what field out of the entry), and a synapse type.   The type specifies
+    the function that normalizes the output of the datapath query into a string or integer.
+
+    Indices can be added and deleted asynchronously from the indexing thread via CryotankIndexer.addIndex and
+    CryotankIndexer.delIndex.
+
+    Indexes can be queried with normValuByPropVal, normRecordsByPropVal, rawRecordsByPropVal.
+
+    To harmonize with LMDB requirements, writing only occurs on the worker thread, while reading indices takes place in
+    the caller's thread.  Both reading and writing index metadata (that is, information about which indices are
+    running) take place on the worker's thread.
+
+    N.B. The indexer cannot detect when a type has changed from underneath itself.   Operators must explicitly delete
+    and re-add the index to avoid mixed normalized data.
+    '''
+    MAX_WAIT_S = 10
+
+    def __init__(self, cryotank):
+        '''
+        Create an indexer.
+
+        Args:
+            cryotank: the cryotank to index
+        Returns:
+            None
+        '''
+        self.cryotank = cryotank
+        ebus = cryotank
+        self._going_down = False
+        self._worker = threading.Thread(target=self._workerloop, name='CryoTankIndexer')
+        path = s_common.gendir(cryotank.path, 'cryo_index.lmdb')
+        cryotank_map_size = cryotank.lmdb.info()['map_size']
+        self._dbenv = lmdb.open(path, writemap=True, metasync=False, max_readers=8, max_dbs=4,
+                                map_size=cryotank_map_size)
+        # iid, v -> offset table
+        self._idxtbl = self._dbenv.open_db(b'indices', dupsort=True)
+        # offset, iid -> normalized prop
+        self._normtbl = self._dbenv.open_db(b'norms')
+        self._to_delete = {}  # type: Dict[str, int]
+        self._workq = s_queue.Queue()
+        # A dict of propname -> version, type, datapath dict
+        self._meta = _IndexMeta(self._dbenv)
+        self._next_offset = self._meta.lowestProgress()
+        self._chunk_sz = 1000  # < How many records to read at a time
+        self._remove_chunk_sz = 1000  # < How many index entries to remove at a time
+        ebus.on('cryotank:puts', self._onData)
+
+        self._worker.start()
+
+        def _onfini():
+            self._going_down = True
+            self._workq.done()
+            self._worker.join(self.MAX_WAIT_S)
+
+        ebus.onfini(_onfini)
+
+    def _onData(self, unused):
+        '''
+        Wake up the worker if he already doesn't have a reason to be awake
+        '''
+        if 0 == len(self._workq):
```

Comment:
since this event handler is going to be called quite a bit, I did some microbenchmarking to see how fast doing the identify comparison against 0 (taking advantage of singletons for small ints) was but the difference was negligible

Reply:
Ideally, our Queue class would implement a __bool__ function so I could just do `if not self._workq`